xdsl/interpreter.py

```diff
@@ -0,0 +1,288 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import (IO, Any, Callable, Generator, Iterable, TypeAlias, TypeVar,
+                    ParamSpec)
+
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import OperationInvT, SSAValue, Operation
+from xdsl.utils.exceptions import InterpretationError
+
+
+@dataclass
+class InterpreterFunctions:
+    """
+    Hold the Python implementations for Operations. Users should
+    subclass this class, and define the functions to run during interpretation.
+    For example:
+
+    ``` python
+    @register_impls
+    class ArithFunctions(InterpreterFunctions):
+
+        @impl(arith.Addi)
+        def run_addi(self, interpreter: Interpreter, op: arith.Addi, 
+                     args: tuple[Any, ...]) -> tuple[Any, ...]:
+            lhs, rhs = args
+            return lhs + rhs, 
+    ```
+
+    The interpreter will take care of fetching the Python values associated 
+    with the operand SSAValues, and setting the return values to the 
+    appropriate OpResults.
+
+    To override the definition of an operation implementation, subclass the
+    class to override, and redefine the functions, annotating them with 
+    `@impl`.
+
+    ``` python
+    @register_impls
+    class DebugArithFunctions(ArithFunctions):
+
+        @impl(arith.Addi)
+        def run_addi(self, interpreter: Interpreter, op: arith.Addi, 
+                     args: tuple[Any, ...]) -> tuple[Any, ...]:
+            lhs, rhs = args
+            print(lhs, rhs, lhs + rhs)
+            return lhs + rhs, 
+    ```
+    """
+
+    @classmethod
+    def impls(
+        cls
+    ) -> Iterable[tuple[type[Operation], OpImpl[InterpreterFunctions,
+                                                Operation]]]:
+        ...
```

Comment:
Nit: Please raise an exception here instead of `...`, as this will fail silently.

Reply:
Good idea, implemented